napari-workshops/notebooks/spot_detection_functions.md

```diff
@@ -0,0 +1,408 @@
+
+```{code-cell} ipython3
+@magicgui(auto_call=True,
+            sigma={"widget_type": "FloatSlider", "min": 0, "max": 20}
+         )
+def gaussian_high_pass(image: "napari.types.ImageData", sigma: float = 2)->"napari.types.ImageData":
+    """Apply a gaussian high pass filter to an image.
+
+    Parameters
+    ----------
+    image : np.ndarray
+        The image to be filtered.
+    sigma : float
+        The sigma (width) of the gaussian filter to be applied.
+        The default value is 2.
+    
+    Returns
+    -------
+    high_passed_im : np.ndarray
+        The image with the high pass filter applied
+    """
+    low_pass = ndi.gaussian_filter(image, sigma)
+    high_passed_im = (image - low_pass).clip(0)
+    
+    return high_passed_im
+```
+
+```{code-cell} ipython3
+viewer.window.add_dock_widget(gaussian_high_pass)
+```
+
+```{code-cell} ipython3
+nbscreenshot(viewer)
+```
+
+Now you can play with the slider until you get the effect you want in the GUI and then return the value:
+
+```{code-cell} ipython3
+gaussian_high_pass.sigma.value
+```
+
+Or you can *set the value*:
+
+```{code-cell} ipython3
+gaussian_high_pass.sigma.value = 3
+```
+
+```{code-cell} ipython3
+nbscreenshot(viewer)
+```
+
+## A more complex example
+
+Finally, lets make a widget for the whole workflow as a function. We will need to write a function
+and then properly annotate it such that `magicgui` can generate the widgets. This time we are also
+starting with image layer (data), but then we want a Points layer with points. We could again return 
+just the layer data using `napari.types.PointsData`. But lets get a nicer Points layer instead, so 
+we will return a LayerDataTuple.  
+
+If `detect_spots()` returns a `LayerDataTuple`, napari will add a *new layer* to
+the viewer using the data in the `LayerDataTuple`. Briefly:
+- The layer data tuple should be: `(layer_data, layer_metadata, layer_type)`
+- `layer_data`: the data to be displayed in the new layer (i.e., the points
+      coordinates)
+- `layer_metadata`: the display options for the layer stored as a
+      dictionary. Some options to consider: `symbol`, `size`, `face_color`
+- `layer_type`: the name of the layer type as a string—in this case `'Points'`  
+
+For more information on using the `LayerDataTuple` type, please see [the documentation](https://napari.org/stable/guides/magicgui.html#returning-napari-types-layerdatatuple).
+
+Also let's change the `image` argument type hint to `ImageLayer` so that we can access more
+properties if we'd like or be able to more easily set the value programmatically.
+
+```{code-cell} ipython3
+# again lets remove the previous widget
+viewer.window.remove_dock_widget("all")
+```
+
+```{code-cell} ipython3
+import numpy as np
+from skimage.feature import blob_log
+
+@magicgui
+def detect_spots(
+    image: "napari.layers.Image",
+    high_pass_sigma: float = 2,
+    spot_threshold: float = 0.2,
+    blob_sigma: float = 2
+    )->"napari.types.LayerDataTuple":
```

Comment:
Look either follow black or PEP8, not this weird mix 😂 (ie this should be de-dented, or you could go 8 spaces for everything (my preference for line continuations)